### PR TITLE
fix(codegen): raise post-verification invariants as InternalError

### PR DESCRIPTION
**Pull request by Hzfengsy**
## What

`src/codegen` and `src/backend` run after the whole pass pipeline, so an invariant that fails there is a compiler bug. **66 of them were raised with `CHECK`**, which throws `pypto::ValueError` and reaches the user as an ordinary Python `ValueError` — indistinguishable from a mistake in their own kernel, with no IR location and no C++ frames to attach to a bug report.

The clearest case said so outright: `pto_codegen.cpp` emitted `"Internal error: missing GM slot buffer offset"` through a macro that throws `ValueError`.

By layer, `src/ir/transforms` sits at roughly 4:1 internal-to-user checks (rule followed); these two trees were near 1:1.

## Changes

**Reclassified 66 checks** to `INTERNAL_CHECK` / `INTERNAL_CHECK_SPAN`:

- The GM slot-offset lookup (`pto_codegen.cpp`) — no `Span` in scope, so the plain form.
- The two `ForStmt`/`WhileStmt` `iter_args_`-vs-`return_vars_` size checks. Six sites in the tree test this predicate; these were the only two not already `INTERNAL_CHECK_SPAN`, and each sits sandwiched between `INTERNAL_CHECK_SPAN` calls on the same node.
- Both duplicated `array.get_element` arity checks, which restate what `DeduceArrayGetElementType` already enforced at construction.
- The arity sweep across `pto_ops_*.cpp` and `src/codegen`, including the shared `CheckArity` helper that fans out to 19 call sites.

Messages are preserved verbatim. **65 of the 66 now also carry `op->span_`**, so they gain the IR source location they previously lacked. A redundant `CHECK`/`INTERNAL_CHECK` pair on the same value in `make_acc_codegen` is collapsed.

**Five span-safety fixes.** `pto_control_flow_codegen.cpp` wrote `INTERNAL_CHECK_SPAN(op != nullptr, op->span_)` in five places — the span argument dereferences the very pointer the check guards, so on failure it segfaults instead of raising. `logging.h:639-646` documents this contract explicitly. These are the only five instances in the repo.

**A regression lint.** These sites cannot be reached from Python by construction — arity and operand types are settled by the op registry at IR-construction time, and `Call`/`ForStmt` have no Python bindings — so no runtime test can hold the classification in place. `tests/lint/check_emitter_check_classification.py` rejects a `CHECK` in either tree whose message says "Internal error" (rule A) or whose predicate tests a call's argument count (rule B), wired into `.pre-commit-config.yaml` beside the sibling `check_*` hooks.

**Docs**: a classification table in `docs/{en,zh}/dev/02-error-handling.md`, and one clause extended in `.claude/rules/error-checking.md` (rules budget unchanged at 2321/2500).

## Deliberately left alone

| | Why |
| --- | --- |
| `l0_tile_chooser.cpp` — 17 `CHECK`s | Load-bearing exception-as-control-flow. `auto_tile_matmul_l0_pass.cpp:1094` catches `pypto::ValueError` *specifically* to emit perf hint PH-AT-005 and skip the matmul. `InternalError` is a **sibling** of `ValueError`, not a subclass, so converting these turns a graceful skip into an uncaught abort. Seven tests pin the messages; one pins the contract end-to-end. |
| `tensor_op_codegen.cpp` `init_value` checks | Genuine user kwarg with no upstream verifier; every message names a remedy. |
| `pto_codegen.cpp` 4-bit dtype checks | Genuine unsupported dtype × backend combination. |
| Post-`As<T>()` type re-checks (~45) and enum-tag decodes (~10) | Same defect class, but each needs individual judgment and several sit next to live `ValueError` tests. The new lint does not flag them. |

## Verification

| Check | Result |
| --- | --- |
| Build (12 touched files force-recompiled) | clean |
| `tests/ut/` | **9978 passed, 3 skipped, 0 failed** |
| `pre-commit run --all-files` | all 18 hooks pass |
| Lint negative check | **64** violations against the parent commit (1 rule A + 63 rule B), **0** after |

**Zero test delta**, as expected: no test in the repo asserted on any converted message, and neither tree contains a `try`/`catch`.

## Notes for reviewers

- **The lint found a site the audit missed.** `orchestration_codegen.cpp:2400` — a `CHECK_SPAN` on `tensor.read`'s arity, which the registry pins at exactly 2. That is why the pre-fix count is 64 rather than 63.
- **The scanner reads one whole macro statement**, not a line window. That matters: `pto_ops_crosscore.cpp` and `scope_outline_utils.h` both put an `INTERNAL_CHECK_SPAN` whose message legitimately says "Internal error" directly below a clean `CHECK`, and a window-based implementation reports the `CHECK` for its neighbour's message. `tests/ut/tools/test_check_emitter_check_classification.py` pins that plus the quoting edge cases (char literals, escaped quotes) — 16 tests.
- **Behavioural caveat**: `pypto.InternalError` subclasses `RuntimeError`, not `ValueError`, so `except ValueError` around `.lower()` no longer swallows these. That is the intended effect, and unreachable in normal use.

### CODE REVIEW

File: tests/lint/check_emitter_check_classification.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    i, n = 0, len(text)
+    while i < n:
+        ch = text[i]
+        nxt = text[i + 1] if i + 1 < n else ""
+        if ch == "/" and nxt == "/":
+            i = _blank_line_comment(text, out, i, n)
+        elif ch == "/" and nxt == "*":
+            i = _blank_block_comment(text, out, i, n)
+        elif ch == '"':
+            i, start = _blank_quoted(text, out, i, n, '"')
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/lint/check_emitter_check_classification.py]: Confirmed and fixed in be7d0c90.

Reproduced exactly as described — the masker treated every `"` as opening an ordinary literal, so `R"(", dtype="opaque", count=)"` (the real one at `src/codegen/distributed/distributed_codegen.cpp:600`) desynchronized it and an arity `CHECK` placed after it returned no violation.

`_blank_comments_and_strings` now dispatches raw literals to their own masking pass: `_raw_prefix_start` recognizes `R`, `LR`, `uR`, `UR` and `u8R` (rejecting an identifier that merely ends in `R`), and `_blank_raw_string` masks through the delimiter-specific `)delim"` terminator.

Seven regression cases in `tests/ut/tools/test_check_emitter_check_classification.py` cover both directions — a check *after* a raw literal is now seen, and arity text *inside* one is not reported — plus custom delimiters, each encoding prefix, and an unterminated literal (which must not hang).

Worth stating plainly: no violation was actually being hidden in the tree. The gate reported zero before this fix and still does, so this was a latent false negative rather than a missed conversion.

File: tests/lint/check_emitter_check_classification.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            strings.append((start, min(i - 1, n)))
+        elif ch == "'":
+            i, _ = _blank_quoted(text, out, i, n, "'")
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/lint/check_emitter_check_classification.py:210]: Confirmed and fixed in be7d0c90.

Reproduced exactly as described — the masker treated every `"` as opening an ordinary literal, so `R"(", dtype="opaque", count=)"` (the real one at `src/codegen/distributed/distributed_codegen.cpp:600`) desynchronized it and an arity `CHECK` placed after it returned no violation.

`_blank_comments_and_strings` now dispatches raw literals to their own masking pass: `_raw_prefix_start` recognizes `R`, `LR`, `uR`, `UR` and `u8R` (rejecting an identifier that merely ends in `R`), and `_blank_raw_string` masks through the delimiter-specific `)delim"` terminator.

Seven regression cases in `tests/ut/tools/test_check_emitter_check_classification.py` cover both directions — a check *after* a raw literal is now seen, and arity text *inside* one is not reported — plus custom delimiters, each encoding prefix, and an unterminated literal (which must not hang).

Worth stating plainly: no violation was actually being hidden in the tree. The gate reported zero before this fix and still does, so this was a latent false negative rather than a missed conversion.

File: tests/lint/check_emitter_check_classification.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+# Sites that are genuinely user-facing despite matching a rule. Each entry is
+# "<repo-relative path>:<line>" and needs a comment saying why the user can actually reach it.
+ALLOWLIST: frozenset[str] = frozenset()
+
+# ``CHECK(`` / ``CHECK_SPAN(`` but never ``INTERNAL_CHECK(`` / ``INTERNAL_CHECK_SPAN(``.
+_MACRO_RE = re.compile(r"(?<![A-Za-z0-9_])(CHECK|CHECK_SPAN)\s*\(")
+
+# ``op->args_.size()`` / ``call->args_.empty()`` and friends.
+_ARITY_RE = re.compile(r"\bargs_\s*\.\s*(?:size|empty)\s*\(")
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/lint/check_emitter_check_classification.py:62]: Partly right, and fixed where it was — but the premise does not hold for two of the three documents.

**Correct, and now fixed:** `.claude/rules/error-checking.md:7` did overclaim, on both counts. It said argument-count *and* post-`As<T>()` checks "are always internal — [the lint] enforces this". Neither was true: the lint never detected post-`As<T>()` checks, and 34 of them in these two directories are still `CHECK`. That sentence is now scoped to what the gate actually does.

**Not correct:** the enforcement sentences in `docs/en/dev/02-error-handling.md:265` and `docs/zh/dev/02-error-handling.md:257` already named exactly the two rules the lint implements — "a `CHECK` ... whose message says \"Internal error\", and a `CHECK` on a call's argument count". Neither claimed post-`As<T>()` coverage, so there was nothing to narrow. What they did lack was a statement that their table is policy rather than a description of the current tree; both now say so and give the 34-site count.

**On the other option offered — adding post-`As<T>()` detection:** rejected deliberately. Those 34 sites need per-site judgment, and several sit beside tests that assert `ValueError`. Converting them was scoped out of this change as a follow-up, so enforcing the rule now would fail the gate on work that was deferred on purpose rather than overlooked. Narrowing the claim keeps the gate honest about its own coverage, which is what both of you asked for.

Fixed in be7d0c90.

File: .claude/rules/error-checking.md [hunk 4]
```diff
@@ -4,7 +4,7 @@
...
 
-**Inside passes**: passes operate on IR that earlier passes have already verified. A failed invariant inside a pass therefore almost always indicates a **compiler bug** — use `INTERNAL_CHECK_SPAN`, not `CHECK`. Reserve `CHECK` / `CHECK_SPAN` for documented user-facing limitations (e.g. "feature X not yet supported — use Y").
+**Inside passes, codegen, and backend emitters**: these operate on IR that earlier passes have already verified. A failed invariant there therefore almost always indicates a **compiler bug** — use `INTERNAL_CHECK_SPAN`, not `CHECK`. Reserve `CHECK` / `CHECK_SPAN` for documented user-facing limitations (e.g. "feature X not yet supported — use Y"). In `src/codegen` / `src/backend` specifically, argument-count and post-`As<T>()` checks are always internal — `tests/lint/check_emitter_check_classification.py` enforces this.
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply .claude/rules/error-checking.md]: Partly right, and fixed where it was — but the premise does not hold for two of the three documents.

**Correct, and now fixed:** `.claude/rules/error-checking.md:7` did overclaim, on both counts. It said argument-count *and* post-`As<T>()` checks "are always internal — [the lint] enforces this". Neither was true: the lint never detected post-`As<T>()` checks, and 34 of them in these two directories are still `CHECK`. That sentence is now scoped to what the gate actually does.

**Not correct:** the enforcement sentences in `docs/en/dev/02-error-handling.md:265` and `docs/zh/dev/02-error-handling.md:257` already named exactly the two rules the lint implements — "a `CHECK` ... whose message says \"Internal error\", and a `CHECK` on a call's argument count". Neither claimed post-`As<T>()` coverage, so there was nothing to narrow. What they did lack was a statement that their table is policy rather than a description of the current tree; both now say so and give the 34-site count.

**On the other option offered — adding post-`As<T>()` detection:** rejected deliberately. Those 34 sites need per-site judgment, and several sit beside tests that assert `ValueError`. Converting them was scoped out of this change as a follow-up, so enforcing the rule now would fail the gate on work that was deferred on purpose rather than overlooked. Narrowing the claim keeps the gate honest about its own coverage, which is what both of you asked for.

Fixed in be7d0c90.

**Review by lyfne123 — APPROVED** [review]
